### ISSUE TITLE
JFrog CLI uploading with wrong module name

### ISSUE THREAD

**Issue by Flash1212** [jfrog/jfrog-cli#622]
**Describe the bug**
A clear and concise description of what the bug is:
JFrog-CLI command `./jfrog rt npm-publish` publishes artifact to artifactory with incorrect module-ver when package "cypress:4.0.1" is found in the package.json dependency list.

**To Reproduce**

1. Create npm package
2. Include "cypress:4.0.1" as dependency
3. ./jfrog rt c artifactory --url=${server} --apikey=${APIKEY} --interactive=false"
4. ./jfrog rt npmi --build-name=${build_name} --build-number=${env.BUILD_NUMBER}"
5. npm run build
6. npm test
7. ./jfrog rt npm-publish --build-name=${build_name} --build-number=${env.BUILD_NUMBER}"

**Expected behavior**

1. Install all dependencies
2. build
3. test
4. pack with "package.json" specifications: "nodejs-hello-world-1.0.0"
5. publish tarball:  "nodejs-hello-world-1.0.0.tgz"

**Screenshots**
If applicable, add screenshots to help explain your problem:

```
- [Info] [Thread 2] Uploading artifact: /var/lib/jenkins/workspace/kbench-CICD_hello-npm_develop/nodejs-hello-world-1.0.0.tgz
- [Debug] Sending HTTP PUT request to:
- https://<server>/artifactory/npm/cypress/-/cypress-4.1.0.tgz;build.name=Sandbox%3A%3Aworkbench%3A%3ACICD%3A%3Ahello-npm%3A%3Adevelop;build.number=86;build.timestamp=1583611046597
-  [Debug] [Thread 2] Artifactory response: 201 Created
- [Debug] Uploaded 1 artifacts.
- [Debug] Successfully deleted the created npm package: /var/lib/jenkins/workspace/kbench-CICD_hello-npm_develop/nodejs-hello-world-1.0.0.tgz
- [Debug] Saving npm package artifact build info data.
- [Debug] Creating temp build file at: /var/lib/jenkins/workspace/kbench-CICD_hello-npm_develop/.jfrog/projects/tmp/jfrog/builds/RGV2T3BzOjpTYW5kYm94Ojp3b3JrYmVuY2gtREo6OkNJQ0Q6OmhlbGxvLW5wbTo6ZGV2ZWxvcF84Ng==/partials
- [Info] npm publish finished successfully.[Pipeline] echoPack n Publish Artifacts: nodejs-hello-world-1.0.0.tgz
```

**Versions**
- JFrog CLI version: 1.34.
- JFrog CLI operating system: Linux <various containers>
- Artifactory Version: Artifactory Enterprise 6.9.1 rev 60901900

**Additional context**
Add any other context about the problem here:
Just the mere existence of this dependency causes this issue. I've tried different versions of it with same results. I bring this here, because this is not an issue when running "npm pack" manually.  
```
[Pipeline] sh (hide)
+ npm pack
[Pipeline] echo
nodejs-hello-world-1.0.0.tgz
```

Also posted here: https://www.jfrog.com/jira/browse/RTFACT-21486?jql=text%20~%20%22jfrog%20cli%22

**Comment by Flash1212** [comment]
Ok, so I dug through the code for jfrog-cli, added a bunch of debugs to read out sections and came across this...

String Has Suffix 'package.json':  package/.cache/Cypress/4.1.0/Cypress/resources/app/package.json
This ties to this [function](https://github.com/jfrog/jfrog-cli/blob/1c26ebbd4474f2eaea4999ba7aab9e72683067f9/artifactory/commands/npm/publish.go#L281). Note on line 302 this condition 
`if strings.HasSuffix(hdr.Name, "package.json") `
will match the first byte returned matching that string. In my circumstance cypress creates a .cache which happens to contain its own "package.json" as shown above. 

After changing it to:
`if strings.HasSuffix(hdr.Name, "package/package.json")`
 It skipped over the .cache in favor of the primary package.json. Does anyone see any reason why this shouldn't be changed?

Would it break some other function that you can think of?

**Comment by gfvirga** [comment]
Hello,

Could this be related?

```
$:~/DC$ export JFROG_CLI_LOG_LEVEL=DEBUG
$:~/DC$ jfrog rt npmp "nxg-npm-dev" --url="https://bar.example.com/artifactory" --user XXXXX --password XXXX
[Warn] You are using a deprecated syntax of the "npmp" command.
        To use the new syntax, the command expects the details of the Artifactory server and repositories to be pre-configured.
        To create this configuration, run the following command from the root directory of the project:
        $ jfrog rt npmc
        This will create the configuration inside the .jfrog directory under the root directory of the project.
        The new command syntax looks very similar to the npm CLI command i.e.:
        $ jfrog rt npmp [npm args and option] --build-name=*BUILD_NAME* --build-number=*BUILD_NUMBER*
[Info] Running npm Publish
[Debug] Preparing prerequisites.
[Debug] Sending usage info...
[Debug] Using npm executable: /usr/bin/npm
[Debug] Working directory set to: /home/user/DC
[Debug] Reading Package Json.
(...)
dc-0.0.15.tgz
[Debug] Created npm package at /home/user/DC/dc-0.0.15.tgz
[Debug] Deploying npm package.
[Debug] Extracting info from npm package: /home/user/DC/dc-0.0.15.tgz
[Info] [Thread 2] Uploading artifact: /home/user/DC/dc-0.0.15.tgz

[Debug] Sending HTTP PUT request to: https://bar.example.com/artifactory/private-npm-dev//-/-.tgz;

[Debug] Artifactory response: 202 Accepted
[Debug] [Thread 2]  Artifactory response: 201 Created 
[Debug] Uploaded 1 artifacts.
[Debug] Successfully deleted the created npm package: /home/user/DC/dc-0.0.15.tgz
[Info] npm publish finished successfully.

```
The command is pushing unnamed file to private-npm-dev//-/-.tgz;
Is there a version that this currently work?

**Comment by Flash1212** [comment]
gfvirga
What's in your package.json file for name and version? It looks like it couldn't get a name at all.

It works for me in ver 1.34. What version are you running?

**Comment by yahavi** [comment]
@Flash1212 thanks for reporting this.
I'm not familiar with Cypress. Please let me know what you think:
* Cypress uses a cache folder. As described [here](https://docs.cypress.io/guides/getting-started/installing-cypress.html#Binary-cache), in Linux it is under this folder: `~/.cache/Cypress`.
* Your build contains somehow the cache. My guess is - You are running builds in the HOME directory.
* `npm pack` packs the cache folder along with your application. It makes a huge tar file.
* `jfrog rt npmp` publishes many package.json files inside the tar and parses one of them as the module name, accidentally.

So my question is - Do you really need the Cypress cache in your tar file? 
Isn't it be more reasonable to change the cache folder to somewhere else?

**Comment by Flash1212** [comment]
Hey @yahavi, 
Thanks for the reply. I'm already "working around" this issue by deleting the cache folder. 

I don't know that I agree that it's "more reasonable" to have to do this. What if there are other cache folders/files that are created that I don't know about and come before 'Package.json'? Then am I responsible for finding these in advance of running jfrog-cli to make sure it works correctly?

This only became an issue when we decided to utilize the jfrog-cli over npm alone. This doesn't happen when running NPM directly. So, you tell me if this isn't something worth changing?

**Comment by yahavi** [comment]
@Flash1212 as I said, you'll get a huge tar when you run `npm pack`. Do you really need the entire Cypress cache in your Tar?
You also don't have to delete the cache folder - Just set the environment variable
`CYPRESS_CACHE_FOLDER=/tmp/cypress_cache`.

Please let me know if it helped.

**Comment by Flash1212** [comment]
@yahavi No, as I stated above, I'm deleting the .cache in my Jenkins build. This .cache is generated after running ./jfrog rt npmi. I have no way in advance to know that this directory will be created in the home directory (or anywhere else) which on occasion we choose to preserve. Cypress is just the module that's doing it this time. My point is what if another module say "foo" makes a .cache or a .foo directory. Nothing's stopping jfrog-cli from matching a potential Package.json inside that one after tarballing the directory.

I may not be conveying my thoughts clearly here. I do not want Cypress or any other cache. I just don't know in advance if a module will create one. I don't write the source of applications being developed.

**Comment by yahavi** [comment]
Thanks, @Flash1212. 
We are working on a fix.

**Comment by yahavi** [comment]
@Flash1212
JFrog CLI v1.35.1 is released. With that version, a Tarball with multi package.json files should work by reading the module info of `pacakge/package.json`, as you suggested.
Feel free to upgrade.

Please note also that except for a very specific reason, you should not pack the Cypress caches in your npm application. This is easy as setting the environment variable `CYPRESS_CACHE_FOLDER` to somewhere else outside the workspace.

We'll also appreciate your feedback for that.

**Comment by Flash1212** [comment]
Hello Yahavi, thank you for the feed back. We appreciate the prompt response to this concern. I'll likely be looking to set a global cache location in a place not packed. 

Thank you.

**Comment by Flash1212** [comment]
@yahavi Also, please note that a ticket was opened [here](https://www.jfrog.com/jira/browse/RTFACT-21486?jql=text%20~%20%22jfrog%20cli%22).

**Comment by eyalbe4** [comment]
@Flash1212,
@yahavi 's fix for this issue released and is included as part of version 1.35.1.
Please let us know if this issue is resolved for you.

**Comment by Flash1212** [comment]
All good now. Thank you.
